### ISSUE TITLE
Sending customMessage() with attachment payload via DM when no previous conversation

### ISSUE THREAD

**Issue by ehammersley** [slackhq/hubot-slack#256]
Very similar to #204, which was resolved via #168 in cases where adapter.send is in use.  Based on what I can understand it checks for an existing IM.id with the user, and if one doesn't exist it opens a new IM.

However, I had a need for a formatted attachment to be sent via DM.  I ran across and followed the customMessage() method outlined in the comments of #170, which basically involves creating the msgData necessary for an attachment and passing it as data to .customMessage.  However, this only works if a previous conversation existed between the bot and user.  If no previous IM id exists it silently fails.

Is there a way to take the changes added to adapter.send in the #168 merge and apply that to the customMessage() method in the adapter?  Is this even the best way to do it?  Certainly open to suggestions on my approach as well.

Here's an idea of how I'm building the data and sending it.  It works great, so long as it's an existing private DM, otherwise nothing.

```
msgData = {
          channel: msg.envelope.user.name
          attachments: [
            {
              fallback: "Fallback text here.",
              color: "danger",
              title: "Title text here"
              text: "Main body text"
              mrkdwn_in: ["text"]
            }
          ]
        }
        robot.adapter.customMessage msgData
```

**Comment by evansolomon** [comment]
This sounds like it makes sense. Seems like the right approach would be to create a DM instead of returning [here](https://github.com/jtatum/hubot-slack/blob/6a3ceefff159115c61819851b383c8e2406405b6/src/slack.coffee#L259-L260). Is that right?

**Comment by ehammersley** [comment]
That's what I'm thinking, yes.  Something is done very similar to this in master that resolves the issue for adapter.send.  However, the structure of customMessage would make the approach a little different I think.

Here's how it's handled in master for .send.
```
channel = @client.getChannelGroupOrDMByName envelope.room
channel = @client.getChannelGroupOrDMByID(envelope.room) unless channel

if not channel and @client.getUserByName(envelope.room)
  user_id = @client.getUserByName(envelope.room).id
  @client.openDM user_id, =>
    this.send envelope, messages...
  return
```
In customMessage the eval is done on channel, with the return, prior to the message being formatted.  I wonder if one could use the above fix for .send, remove the return after the channel eval, and add a conditional send to the end, checking for the DM and creating if necessary, or posting the message if a regular channel?

I'm happy to test a proposed solution.  I'm not sure I'm good enough of a coder to contribute, but I can give it a whirl if needed.

**Comment by evansolomon** [comment]
> I can give it a whirl if needed

That would be awesome

**Comment by ehammersley** [comment]
Here's my attempt, which tests out good in my instances.  It's tricky to test because once you've established that initial DM then you've cornered yourself out of being a valid case.  Answer is lots of test accounts.

I hate to ask but this is my first attempt at anything like this.  Does the below look right?  I have not triggered the pull request yet.  Apologies if the ask for review is abnormal.

https://github.com/ehammersley/hubot-slack/commit/daafc380cf7d72182b6bae06db66cb4c074cfea1

**Comment by ehammersley** [comment]
This one works, but it's not very elegant.  What would be nice is if I could pull the channel ID necessary from the API callback when the DM is created, but I don't know if that's possible.

https://github.com/ehammersley/hubot-slack/commit/4f73b08eb3b62a0f33ba123977f2e9f2db95f161

I know there must be better ways to handle lines 316 and 317 from that callback.  I welcome feedback so I can learn.

**Comment by technicalpickles** [comment]
I've found that this works, and is pretty simple:

```coffeescript
robot.messageRoom room: "slack-user-name-here"
```

It's not very obvious. Actually, reading this a little more closely, it's possible this only works if you already have a DM open with hubot.

**Comment by rajatguptarg** [comment]
@technicalpickles It is not working for me.

I am getting `ERROR TypeError: name.replace is not a function`. How can I fixed that?

**Comment by technicalpickles** [comment]
@rajatguptarg What's the rest the stack trace?

**Comment by DEGoodmanWilson** [comment]
I _believe_ the Slack API has changed so that opening a DM is not necessary for posting the first DM to a user. Does this remain a live issue for y'all?

**Comment by DEGoodmanWilson** [comment]
Closing this as fixed in 4.0. Please reopen if I am mistaken!

**Comment by ehammersley** [comment]
Apologies.  I made the change above in my branch and initiated the pull, then never rebased my production bot without the change.  I meant to test this and never found the time.  I'll try and do that and reopen if necessary.
